Inventory on string metrics

jotak opened the pull request:
(Got all integration tests green on my machine, now checking with travis...)

jotak committed:
Inventory with strings / first draft

jotak committed:
Inventory with strings / first draft (tests)

jotak committed:
Inventory on metrics

- Fix integration tests
- Deleting a root resource <=> push empty datapoint + untag feed

jotak commented:
... need to set org.hawkular.metrics.string-size to -1 in metrics config db...

jmazzitelli commented:
Putting a hold on any changes/PR merges to the agent until we decide what to do about this major refactoring effort going on right now (see PR #302 ). I've got over 200 files changed (many are wholesale moving of large chunks of files) and any thing going in now it going to potentially involve major conflicts. To avoid any git pain, going to wait until that major refactor is complete. Should be soon.

jotak commented:
Ok, anyway, this change must also be coordinated with the MIQ side and other parts (hServices...) so it's not ready to be merged so fast

jotak committed:
Fix url-encoding issue

Do not url-encode in tags

jmazzitelli commented:
The agent refactor is complete and merged into master. As expected, this PR is affected. Major conflict resolution required.

jotak committed:
Java-agent: fix tests after rebasing

jotak commented on hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/service/AgentCoreEngine.java:
```diff
@@ -291,15 +288,6 @@ public void startHawkularAgent(AgentCoreEngineConfiguration newConfiguration) {
-                    // 1. register our feed ID
-                    // 2. connect to the server's feed comm channel
-                    try {
-                        registerFeed(tenantIds);
```
@jmazzitelli : I removed registerFeed as it doesn't have to be explicitly setup now. Except if you think that a feed should have an existence even if it has nothing associated with (no metric type, no resource type, no root resource)... in which case I should rollback this part.

jotak commented on hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/storage/AsyncInventoryStorage.java:
```diff
@@ -708,31 +704,20 @@ private void postInventoryData(InventoryMetric metric, String jsonPayload, Strin
+        MetricDefinition def = metric.toMetricDefinition();
         if (resourceTypes != null) {
             for (Map.Entry<String, Collection<String>> entry : resourceTypes.entrySet()) {
                 // FIXME: make sure "," cannot be used in ids, or choose another character
```
(note for self: need a fix here)

jotak committed:
Gzip string inventory

jotak committed:
Inventory on metrics: chunk data

jotak committed:
Fix rebase conflicts

jotak committed:
Performances: don't chunk when all fits in 1 chunk

(it makes the economy of a System.arrayCopy)

jotak committed:
Inventory chunks: chunks are datapoints in metric

... instead of having 1 metric per chunk

jotak committed:
Bump hawkular-agent-itest-util

jotak committed:
Avoid systematic updates of inventory

- Use list of added/modified/deleted resources after discovery to select relevant root resources
- In consequence, make InventoryListener API more consistent in regard to its usage with storage: merge various inventory events into 1 event / 1 listen function

jotak commented:
@jmazzitelli can you have a look at the last commit? ( https://github.com/hawkular/hawkular-agent/pull/303/commits/6000ce76bb66376445513b63f8fc43b166fd5adc )

Following our discussion yesterday, I made some changes to avoid unnecessary updates. Had to refactor some code.

jotak commented on hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java:
```diff
@@ -16,58 +16,156 @@
-     * Creates an inventory event.
-     *
-     * @param samplingService a service that provides details such as feed ID and endpoint information that helps
+     * Creates a discovery event.
```
Will update javadoc

jotak replied:
done

jmazzitelli commented on hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java:
```diff
@@ -16,58 +16,156 @@
+    }
+
+    /**
+     * @return the map of removed resources during this discovery
```
The javadoc for this and the one above (getRemovedRootResources, getAddedOrModifiedRootResources) seems wrong. These return the root resources of all the resources that were removed (or added/modified). These root resources weren't necessarily added or removed (though they could have been, too), but they are the root resources of the resources that were added or removed. Right?

jotak replied:
Yes, that's correct. I'm doing a javadoc cleaning pass

jotak replied:
done

jmazzitelli commented on hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java:
```diff
@@ -16,58 +16,156 @@
+        return addedOrModified;
+    }
+
+    public List<Resource<L>> getRemoved() {
```
Here you should add javadoc to these two (getAddedOrModified, getRemoved) explaining that these are the actual resources that were added/modified and removed (which is different than the getXYZRootResources methods).

jotak committed:
Fix javadoc

jmazzitelli commented on hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java:
```diff
@@ -16,58 +16,190 @@
+        removedRootResources = new HashMap<>();
+        addedOrModified.forEach(r -> {
+            Resource<L> root = getRootResource(r);
+            addedOrModifiedRootResources.put(root.getID().getIDString(), root);
```
Lines 83 and 84 tell me that "addedOrModifiedRootResources" will contain the roots of all resources that were added or modified. Is this what you intended? So if a resource changes one of its resource config properties (i.e. it is modified) then its root resource will go into this addedOrModifiedRootResources map (in that case the root isn't really modified - its list of descendants didn't change with a new resource added or old resource removed - it is just that one of its descendants (child, grandchild, or lower) got modified - in this case, do you want the root resource to be considered "modified")?

jotak replied:
Yes, this is what is intended. The reason is that the whole root resource subtree is going to be persisted. So we need to separate root resources that have 1 modification (at least) wherever in their subtree = `addedOrModifiedRootResources` map ; and root resources that need to be entirely deleted = `removedRootResources` map.

jotak committed:
Add metric types tags & index

They are necessary to allow client to query for metrics by metric type - this function was unintentionally removed from the ruby client.

jotak committed:
Manage inventory persistence relatively to metrics TTL

Persistence flag: not just a boolean anymore but a timestamp that will be used to determine if a refresh is needed.
The formula takes into account metrics TTL and auto-discovery scan period.

Refactored the inventory structure builder

Add unit test on AsyncInventoryStorage

jotak commented on hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/storage/AsyncInventoryStorage.java:
```diff
@@ -451,188 +378,156 @@ private String getInventoryId(NamedObject no) {
+        resourceManager.getRootResources().forEach(r -> {
+            // Ignore unmodified resources that doesn't need to be refreshed
+            if (addedOrModifiedIds.contains(r.getID()) || needToRefresh(r)) {
+                log.infof("Updating root resource: %s", r.getID().getIDString());
```
@jmazzitelli are you fine with those 3 log.info (also line 444 and 435) that happen every time inventory storage is updated? That can be interesting for tracing http calls, but if you prefer I can put to debug level (or remove).

jmazzitelli replied:
you can move them to debug level

jotak committed:
Remove obsolete FIXME comments

jotak committed:
Change some logs to debug level

jmazzitelli commented:
@metlos took this PR and built upon it and that is now PR #320 which supercedes this PR. This PR will be closed.

jmazzitelli commented:
PR #325 is the one we are going to use - close this one
